components/node/rollup/derive/engine_queue.go

```diff
@@ -389,8 +419,11 @@ func (eq *EngineQueue) logSyncProgress(reason string) {
 // tryUpdateEngine attempts to update the engine with the current forkchoice state of the rollup node,
 // this is a no-op if the nodes already agree on the forkchoice state.
 func (eq *EngineQueue) tryUpdateEngine(ctx context.Context) error {
+	if eq.isEngineSyncing() {
+		eq.log.Warn("Attempting to update forkchoice state while engine is P2P syncing")
+	}
 	fc := eth.ForkchoiceState{
-		HeadBlockHash:      eq.unsafeHead.Hash,
+		HeadBlockHash:      eq.engineSyncTarget.Hash,
```

Comment:
This `tryUpdateEngine` is called when fork choice is needed. 
But can we sure that `engineSyncTarget` is always same with the `unsafeHead` when sync is done?

Reply:
During engine is syncing, when a node receives unsafe payload, node continuously requests geth to update unsafe block using `tryUpdateEngine`. In this case, `eq.unsafeHead` is not changed but `eq.engineSyncTarget` is updated with the newly received unsafe payload. So here we use `eq.engineSyncTarget`. And after engine has finished syncing, a node requests geth to update safe and finalized block using `tryUpdateEngine`. When `eq.engineSyncTarget` and `eq.unsafeHead` is different, it will do not `tryUpdateEngine` thanks to the line 257-260.